src/Pipeline/Atomic.php

```diff
@@ -62,56 +64,82 @@ protected function getConnection()
+
+        // transmit all commands in transaction
         foreach ($commands as $command) {
             $connection->writeRequest($command);
         }
 
+        // execute transaction: EXEC
+        $connection->writeRequest($exec);
+
+        // response to MULTI
+        $response = $connection->readResponse($multi);
+        if ($response instanceof ErrorResponseInterface) {
+            $this->exception($connection, $response); // close connection and throw ServerException
+        }
+
+        // responses to commands (all should be QUEUED result)
+        $ex = null;
         foreach ($commands as $command) {
             $response = $connection->readResponse($command);
 
-            if ($response instanceof ErrorResponseInterface) {
-                $connection->executeCommand($profile->createCommand('discard'));
-                throw new ServerException($response->getMessage());
+            // throw first error as a ServerException
+            if (($response instanceof ErrorResponseInterface) && !$ex) {
+                $ex = new ServerException($response->getMessage());
             }
         }
 
-        $executed = $connection->executeCommand($profile->createCommand('exec'));
+        // response to EXEC
+        $responses = $connection->readResponse($exec);
 
-        if (!isset($executed)) {
-            // TODO: should be throwing a more appropriate exception.
-            throw new ClientException(
-                'The underlying transaction has been aborted by the server.'
-            );
+        // if one of the commands returned an error, only throw after all responses have been received
+        if ($ex) {
+            throw $ex;
         }
 
-        if (count($executed) !== count($commands)) {
+        // if EXEC returned null, the transaction was aborted
+        if (!isset($responses)) {
+            // fake a MultiExec, which is not used in an atomic pipeline
+            $multi_exec = new MultiExec($this->getClient());
+            throw new AbortedMultiExecException($multi_exec, 'The underlying transaction has been aborted by the server.');
+        }
+
+        // if EXEC returned an error, throw ServerException (but don't close connection, all responses
+        // have been read and are accounted for)
+        if ($responses instanceof ErrorResponseInterface) {
+            throw new ServerException($responses->getMessage());
+        }
+
+        if (count($responses) !== count($commands)) {
             $expected = count($commands);
-            $received = count($executed);
+            $received = count($responses);
 
             throw new ClientException(
                 "Invalid number of responses [expected $expected, received $received]."
             );
         }
 
-        $responses = array();
-        $sizeOfPipe = count($commands);
         $exceptions = $this->throwServerExceptions();
 
-        for ($i = 0; $i < $sizeOfPipe; ++$i) {
-            $command = $commands->dequeue();
```

Comment:
Is it really necessary to change this?

This was meant to free memory as soon as possible by discarding useless instances of `Predis\Command\Command` stored in `$commands`: basically when a `$response` is parsed by the associated `$command` we don't need to keep `$command` in memory, so doing a `SplQueue::dequeue()` removes the only reference for that instance that can be reclaimed by the GC at the next iteration). When you have some hefty pipelines this actually makes a difference in terms of memory usage.

Even by going with your change, you are still not emptying the `SplQueue` (which is kept in reference by the underlying implementation of `Predis\Pipeline\Pipeline`). Truth to be told `SplQueue` should probably be emptyed even when exception occurs but this is not done even in my original code, this should be investigated but it's not the scope of this PR as it also involves the base class.

Reply:
I wasn't aware emptying the queue was a design goal.  I'll update the branch with that in mind.

Reply:
I've reverted this change.  The queue is drained as the responses are processed.